src/System.IO.FileSystem/src/System/IO/FindEnumerable.Windows.cs

```diff
@@ -0,0 +1,252 @@
+        private void Initialize()
+        {
+            _currentPath = _originalFullPath;
+            _buffer = ArrayPool<byte>.Shared.Rent(4096);
+            _pinnedBuffer = GCHandle.Alloc(_buffer, GCHandleType.Pinned);
+            if (_recursive)
+                _pending = new Queue<(IntPtr, string)>();
+            _directoryHandle = CreateDirectoryHandle(_originalFullPath);
+        }
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        public TResult Current => _current;
+
+        object IEnumerator.Current => Current;
+
+        public bool MoveNext()
+        {
+            if (_lastEntryFound)
+                return false;
+
+            lock (_lock)
+            {
+                if (_lastEntryFound)
+                    return false;
+
+                RawFindData findData = default;
+                do
+                {
+                    FindNextFile();
+                    if (!_lastEntryFound && _info != null)
+                    {
+                        // If needed, stash any subdirectories to process later
+                        if (_recursive && (_info->FileAttributes & FileAttributes.Directory) != 0
+                            && !PathHelpers.IsDotOrDotDot(_info->FileName))
+                        {
+                            string subDirectory = PathHelpers.CombineNoChecks(_currentPath, _info->FileName);
+                            IntPtr subDirectoryHandle = CreateDirectoryHandle(subDirectory);
+                            try
+                            {
+                                _pending.Enqueue((subDirectoryHandle, subDirectory));
+                            }
+                            catch
+                            {
+                                Interop.Kernel32.CloseHandle(subDirectoryHandle);
+                                throw;
+                            }
+                        }
+
+                        findData = new RawFindData(_info, _currentPath, _originalFullPath, _originalUserPath);
+                    }
+                } while (!_lastEntryFound && !_predicate(ref findData, _state));
+
+                if (!_lastEntryFound)
+                    _current = _transform(ref findData);
+
+                return !_lastEntryFound;
+            }
+        }
+
+        private unsafe void FindNextFile()
+        {
+            Interop.NtDll.FILE_FULL_DIR_INFORMATION* info = _info;
+            if (info != null && info->NextEntryOffset != 0)
+            {
+                // We're already in a buffer and have another entry
+                _info = (Interop.NtDll.FILE_FULL_DIR_INFORMATION*)((byte*)info + info->NextEntryOffset);
+                return;
+            }
+
+            // We need more data
+            if (GetData())
+                _info = (Interop.NtDll.FILE_FULL_DIR_INFORMATION*)_pinnedBuffer.AddrOfPinnedObject();
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void NoMoreFiles()
```

Comment:
I find this a strange method name.  What does the method do?  It appears to be stating a fact rather than saying what action it will take.

Reply:
Fair enough- that was the error code.  I'll change to DirectoryFinished.